Fix part of #13162: Add schema for AudioUploadHandler

adamhalim opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13162.
2. This PR does the following: Adds schemas for AudioUploadHandler

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
Print statements after hitting the endpoint:
```
------------------------------------
Request url =  http://localhost:8181/createhandler/audioupload/1?filename=file
Handler class name =  AudioUploadHandler
handler_args =  {'filename': 'file', 'raw_audio_file': '123'}
Arguments =  ['filename', 'raw_audio_file']
Iterating over arguments...
filename file
raw_audio_file 123
URL path elements =  {'exploration_id': '1'}
Request method =  POST
HANDLER_ARGS_SCHEMAS =   {'POST': {'raw_audio_file': {'schema': {'type': 'basestring'}}, 'filename': {'schema': {'type': 'basestring'}}}}
URL_PATH_ARGS_SCHEMAS = ,  {'exploration_id': {'schema': {'type': 'basestring'}}}
GET_HANDLER_ERROR_RETURN_TYPE html
------------------------------------

```
## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

adamhalim committed:
Fix part of #13162: Add schema for AudioUploadHandler

adamhalim committed:
Merge branch 'oppia:develop' into audio-upload-handler-schema

adamhalim commented:
Not sure if I need to add validators for filename?

adamhalim committed:
Merge branch 'oppia:develop' into audio-upload-handler-schema

sahiljoster32 commented:
Hi @adamhalim, If you are stuck somewhere feel free to ping me or @Nik-09. Thanks!

adamhalim committed:
Merge branch 'develop' into audio-upload-handler-schema

kevintab95 reviewed (approved):
LGTM, thanks @adamhalim!

kevintab95 commented:
Hi @adamhalim, looks like there is a backend test failure. Could you take a look? https://github.com/oppia/oppia/actions/runs/3240293108/jobs/5310749437
Thanks!

vojtechjelinek commented:
@adamhalim Do you need any help with the backend tests?

adamhalim committed:
Add regex validators for exploration_id & filename

adamhalim committed:
Merge branch 'audio-upload-handler-schema' of https://github.com/adamhalim/oppia into audio-upload-handler-schema

adamhalim commented:
> @adamhalim Do you need any help with the backend tests?

@vojtechjelinek

I do have a couple of questions. I took look at the code again, and from my understanding it looked like some validators were missing. I added two validators in my previous commit 24af46c1e9ccd6534fa5077478e126ba89a81ef7, which are some regex validators for the `exploration_id` and `filename`. 

My regex knowledge is pretty bad, so I'm not really sure about the robustness of the `filename` one. Also, what the post method in `AudioUploadHandler` is checking is if the actual file is an mp3 file, not just the file extension. For this to be added to a validator, a validation function has to be executed right? I'm not sure where this function should be written? Also, if the `filename` regex looks good, should this be moved to `constsants.ts` as well? Furthermore, should the regex be connected to `feconf.ACCEPTED_AUDIO_EXTENSIONS`?

In the latest build job https://github.com/oppia/oppia/actions/runs/3297856487/jobs/5439206538, the following 6 tests fail:

1. `test_cannot_upload_audio_with_invalid_exp_id`
2. `test_detect_non_audio_file`
3. `test_detect_non_matching_extensions`
4. `test_guest_can_not_upload`
5. `test_invalid_extension_is_detected`
6. `test_missing_extensions_are_detected`

For `test_cannot_upload_audio_with_invalid_exp_id`, the failure is `AssertionError: 400 != 404`.
- This tests sends a POST request to `/createhandler/audioupload/invalid_exp_id`, which responded with a 404 before I added the regex validation for the exploration_id. Have I done something wrong, or should the test now expect 400?

For `test_detect_non_audio_file` the failure is  `- Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\s]+(\\.(?i)(mp3))$'}) for object cafe.flac`
- If the regex shouldn't be used, I'll just remove it and it's fine. Still, a validation function needs to be written(?), Otherwise, should I just add this to the test's assertion? 

For `test_detect_non_matching_extensions`, the failure is `AssertionError: 'Although the filename extension indicates the file is a flac file, it was not recognized as one. Found mime types:' not found in "Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\\\s]+(\\\\.(?i)(mp3))$'}) for object test.flac"`
- Same as above.

For `test_guest_can_not_upload`, the failure is `AssertionError: 302 != 401`
- Not sure why 302 is being returned?

For `test_invalid_extension_is_detected`, the failure is `- Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\s]+(\\.(?i)(mp3))$'}) for object test.wav`
- Add this to the test's assertion? 

For `test_missing_extensions_are_detected`, the failure is `- Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\s]+(\\.(?i)(mp3))$'}) for object test`
- Same as above

vojtechjelinek commented on core/controllers/voice_artist.py:
```diff
@@ -37,15 +38,46 @@ class AudioUploadHandler(base.BaseHandler):
+        }
+    }
+
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'raw_audio_file': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'filename': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'is_regex_matched',
+                        'regex_pattern': r'[^\s]+(\.(?i)(mp3))$'
+                    }]
+                }
+            }
+        }
+    }
+
```
I think your validators should make sure that all the checks that raise `InvalidInputException` in the `post` method are not needed.

vojtechjelinek commented:
> I do have a couple of questions. I took look at the code again, and from my understanding it looked like some validators were missing. I added two validators in my previous commit [24af46c](https://github.com/oppia/oppia/commit/24af46c1e9ccd6534fa5077478e126ba89a81ef7), which are some regex validators for the `exploration_id` and `filename`.

The validations look good.

> My regex knowledge is pretty bad, so I'm not really sure about the robustness of the `filename` one. Also, what the post method in `AudioUploadHandler` is checking is if the actual file is an mp3 file, not just the file extension. For this to be added to a validator, a validation function has to be executed right? I'm not sure where this function should be written? Also, if the `filename` regex looks good, should this be moved to `constsants.ts` as well? Furthermore, should the regex be connected to `feconf.ACCEPTED_AUDIO_EXTENSIONS`?

I left some comments in your code, can you please explain your regex a bit?

> For `test_cannot_upload_audio_with_invalid_exp_id`, the failure is `AssertionError: 400 != 404`.
> 
>     * This tests sends a POST request to `/createhandler/audioupload/invalid_exp_id`, which responded with a 404 before I added the regex validation for the exploration_id. Have I done something wrong, or should the test now expect 400?

Yep change the test to accept 400.

> For `test_detect_non_audio_file` the failure is `- Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\s]+(\\.(?i)(mp3))$'}) for object cafe.flac`
> 
>     * If the regex shouldn't be used, I'll just remove it and it's fine. Still, a validation function needs to be written(?), Otherwise, should I just add this to the test's assertion?

Yep you might need to change the assert so that we expect the new format of the error.

> For `test_detect_non_matching_extensions`, the failure is `AssertionError: 'Although the filename extension indicates the file is a flac file, it was not recognized as one. Found mime types:' not found in "Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\\\s]+(\\\\.(?i)(mp3))$'}) for object test.flac"`
> 
>     * Same as above.

dtto

> For `test_guest_can_not_upload`, the failure is `AssertionError: 302 != 401`
> 
>     * Not sure why 302 is being returned?

Not sure either, you will need to investigate.

> For `test_invalid_extension_is_detected`, the failure is `- Schema validation for 'filename' failed: Validation failed: is_regex_matched ({'regex_pattern': '[^\\s]+(\\.(?i)(mp3))$'}) for object test.wav`
> 
>     * Add this to the test's assertion?

Yes.

This is a great explanation, if you would like to discuss some of these over a call, please lmk and we can arrange something.

adamhalim committed:
Update resources_test.py after AudioUploadHandler schema

adamhalim committed:
Merge branch 'develop' into audio-upload-handler-schema

adamhalim committed:
test_invalid_extension_is_detected: change error_msg to test.wav
from test.flac. Made mistake in previous commit

adamhalim commented:
@vojtechjelinek 

I have a question regarding the test `test_detect_non_matching_extensions` in `resources_test.py`.

The test sends an MP3 file, with a filename ending in `.flac`, and the test says it's supposed to
```
# Use an accepted audio extension in mismatched_filename
# that differs from the uploaded file's audio type.
```

Looking at `feconf.py`, the only accepted audio extension is `.mp3`.

Shouldn't the test send a file ending in `.mp3` that uses another MIME type?

Another question I have is how to add this check to the validator? What 
I  mean is, how can I run a function on the input?

vojtechjelinek commented:
> @vojtechjelinek
> 
> I have a question regarding the test `test_detect_non_matching_extensions` in `resources_test.py`.
> 
> The test sends an MP3 file, with a filename ending in `.flac`, and the test says it's supposed to
> 
> ```
> # Use an accepted audio extension in mismatched_filename
> # that differs from the uploaded file's audio type.
> ```
> 
> Looking at `feconf.py`, the only accepted audio extension is `.mp3`.
> Shouldn't the test send a file ending in `.mp3` that uses another MIME type?


Yeah, I think it should.

> Another question I have is how to add this check to the validator? What I mean is, how can I run a function on the input?

If I understand what are you asking is should be this https://github.com/oppia/oppia/blob/f74b606e8511cd4296b3c99aad37e53b66cca196/core/controllers/blog_dashboard.py#L147

adamhalim committed:
Fix test_detect_non_matching_extensions
The test is supposed to upload a file with an accepted
file extension, but with a mismatched MIME type. It
previously uploaded an MP3 file ending in .flac, which
is not an accepted file extension. It now instead sends
a FLAC file ending in .mp3.

adamhalim committed:
Add has_matching_audio_extension to schema_utils.py

adamhalim committed:
Add method docstring to has_matching_audio_extension

adamhalim committed:
Change name from has_matching_audio to is_valid_audio_file

adamhalim committed:
Merge branch 'oppia:develop' into audio-upload-handler-schema

adamhalim committed:
Merge branch 'audio-upload-handler-schema' of https://github.com/adamhalim/oppia into audio-upload-handler-schema

adamhalim commented:
@vojtechjelinek
Thanks for pointing me in the right direction, I think I got something that works now. I added the function `is_valid_audio_file()` to  `core/schema_utils.py`.

While going over the tests, I noticed that some of them use mocked behavior to allow non-mp3 files to be allowed, which now isn't possible due to the regex on filename. This explains why flac files were used for some tests. Should I remove this from the file?
https://github.com/oppia/oppia/blob/90c13355c3891026d78717a8eef5059ef914b50f/core/controllers/resources_test.py#L525-L532

I also presume that all input validation from the handler can be removed? But before I do that, I have a question about these lines of code:
https://github.com/oppia/oppia/blob/90c13355c3891026d78717a8eef5059ef914b50f/core/controllers/voice_artist.py#L68-L82
In `is_valid_audio_file()`, I only run `audio = mp3.MP3(tempbuffer)` and skip the extension check, since anything that doesn't end in .mp3 is invalid anyways. Is this OK? Or do we actually want to handle a case where filename.flac is an mp3 file (or flac files in general)?

Lastly I still got the issue with one test failing (same as before), I'll investigate this week and see if I can find the cause.

chris7716 commented:
Hi @vojtechjelinek could you PTAL. It seems @adamhalim got few questions for you. Thanks!

vojtechjelinek commented:
> @vojtechjelinek Thanks for pointing me in the right direction, I think I got something that works now. I added the function `is_valid_audio_file()` to `core/schema_utils.py`.
> 
> While going over the tests, I noticed that some of them use mocked behavior to allow non-mp3 files to be allowed, which now isn't possible due to the regex on filename. This explains why flac files were used for some tests. Should I remove this from the file?
> 
> https://github.com/oppia/oppia/blob/90c13355c3891026d78717a8eef5059ef914b50f/core/controllers/resources_test.py#L525-L532

Yep, feel free to remove it.

> I also presume that all input validation from the handler can be removed? But before I do that, I have a question about these lines of code:
> 
> https://github.com/oppia/oppia/blob/90c13355c3891026d78717a8eef5059ef914b50f/core/controllers/voice_artist.py#L68-L82

> In `is_valid_audio_file()`, I only run `audio = mp3.MP3(tempbuffer)` and skip the extension check, since anything that doesn't end in .mp3 is invalid anyways. Is this OK? Or do we actually want to handle a case where filename.flac is an mp3 file (or flac files in general)?

Yes, I think it is fine to only validate the `mp3`.

> Lastly I still got the issue with one test failing (same as before), I'll investigate this week and see if I can find the cause.

adamhalim committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audio-upload-handler-schema

adamhalim committed:
Removed mock_accepted_audio_extensions from resources_test

adamhalim committed:
Remove input validation from AudioUploadHandler.post()
Validation is done through the validator in the schema

adamhalim commented:
@vojtechjelinek 
I think I finally found why the last test fails: https://github.com/oppia/oppia/blob/f8a8d680c6f95dfc3c63a003484177ac06e4ce21/core/controllers/resources_test.py#L534-L551

Currently, I get the following error: `AssertionError: 302 != 401`. I managed to find where this happens, which is this code right here: https://github.com/oppia/oppia/blob/f8a8d680c6f95dfc3c63a003484177ac06e4ce21/core/controllers/base.py#L756-L783

Running the test, the variable `payload_exists` evaluates to `False`, and `self.GET_HANDLER_ERROR_RETURN_TYPE` is `html`.

`self.payload` is equal to  `<core.controllers.base.RaiseErrorOnGet object at 0x7f4b3d94e730>`, and `self.normalized_payload` is equal to `{'filename': 'cafe.mp3'}`.

From my understanding, since the schema is now implemented and the normalized arguments are used, `payload_exists` will be `False` and the function now redirects (with a 302) instead of responding with 401.

Is this correct behavior? I haven't observed this anywhere else and am not sure if I should just change the test to expect 302, or if the implementation is incorrect.

adamhalim committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audio-upload-handler-schema

adamhalim committed:
Fix linter errors

vojtechjelinek commented:
@adamhalim Sorry, I missed this I will take a look tomorrow.

vojtechjelinek commented:
@adamhalim I think we can do `GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON` in the handler and see if it fixes the issue. If not I would probably need to see what were the values in _base.py_ before and after your changes.

adamhalim committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audio-upload-handler-schema

adamhalim committed:
Fix mypy checks

adamhalim committed:
Set GET_HANDLER_ERROR_RETURN_TYPE to JSON
Fixes a bug where a 302 redirect would be returned
instead of a 401 unauthorized

adamhalim committed:
Fix lint errors

adamhalim committed:
cast raw_audio_file to bytes
This is to pass mypy checks, instead of using if-else
to check type.

adamhalim committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audio-upload-handler-schema

adamhalim committed:
Removed redundant check in is_valid_audio_file
Anything that isn't an mp3-file should get caught
as a mutagen.MutagenError.
Also, if not obj should already catch if
raw_audio_file is empty

adamhalim commented:
@vojtechjelinek 
> @adamhalim I think we can do `GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON` in the handler and see if it fixes the issue. If not I would probably need to see what were the values in _base.py_ before and after your changes.

That fixed the issue, thanks! I also fixed the new mypy changes and pass all tests now.

adamhalim committed:
Merge branch 'develop' into audio-upload-handler-schema

vojtechjelinek commented on core/schema_utils.py:
```diff
@@ -38,6 +40,9 @@
 from core.domain import user_domain
 from extensions.objects.models import objects
 
+import mutagen
+from mutagen import mp3
+
```
```suggestion
import mutagen

```

vojtechjelinek commented on core/schema_utils.py:
```diff
@@ -26,10 +26,12 @@
 import re
 import urllib
 
+
```
Remove this.

vojtechjelinek commented on core/schema_utils.py:
```diff
@@ -791,3 +796,38 @@ def has_unique_subtitled_contents(
+        # .mp3 is the only allowed extension.
+        extension = 'mp3'
+        try:
+            audio = mp3.MP3(tempbuffer)
```
```suggestion
            audio = mutagen.mp3.MP3(tempbuffer)
```

vojtechjelinek commented on core/controllers/voice_artist.py:
```diff
@@ -41,69 +40,63 @@ class AudioUploadHandler(
+        assert self.normalized_request is not None
+
+        filename = self.normalized_payload['filename']
+        # Here we use cast because we want to narrow down the type
+        # of normalized_request['raw_audio_file'] from  str to bytes.
+        raw_audio_file = cast(bytes, self.normalized_request['raw_audio_file'])
```
```suggestion
        raw_audio_file = self.normalized_request['raw_audio_file'].encode('ascii')
```

adamhalim replied:
@vojtechjelinek 
I tried this initially, but doing this returns the same error for several tests: 

`{"error": "'bytes' object has no attribute 'encode'", "status_code": 500}`
caused by
```
File "~/oppia/core/controllers/voice_artist.py", line 92, in post
    raw_audio_file = self.normalized_request['raw_audio_file'].encode('ascii')
AttributeError: 'bytes' object has no attribute 'encode'
```

I think this is because the `raw_audio_file` is sent as bytes, even though the type hint for `normalized_request` is `Dict[str,str]`. In the schema for `raw_audio_file`, the type is `basestring` which means both `str` and `bytes` should be okay, so I'm not sure what the best solution for this would be?

Furthermore, I think tests that should expect a 200 OK (but don't currently), such as https://github.com/oppia/oppia/blob/28c396663b6616e38e42d558d829dad897064f68/core/controllers/resources_test.py#L579-L595 should assert the response. I might as well make these changes to the tests.

vojtechjelinek replied:
Ah I know what is going on, you will need to add the correct typing. @sahiljoster32 Can you help here?

sahiljoster32 replied:
Hi @adamhalim,

From the discussion, it looks to me that `self.normalized_request['raw_audio_file']` should be typed as bytes rather than str.

So, I think you should define a typedDict for`self.normalized_request`. like we are doing in other files also https://github.com/oppia/oppia/blob/e915b3dccfcc629800b6fa99ccc2817e4d7ae562/core/controllers/editor.py#L94

sahiljoster32 replied:
Ok, for your case it goes something like below:

```python
class AudioUploadHandlerNormalizedRequestDict(TypedDict):
    """Dict representation of AudioUploadHandler's
    normalized_request dictionary.
    """

    raw_audio_file: bytes


```

and pass that `AudioUploadHandlerNormalizedRequestDict` type as a second type argument in basehandler `BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict] ` like below:
```python
class AudioUploadHandler(
    base.BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict]
):
```

adamhalim replied:
Hey @sahiljoster32 , thanks for your comments.

I just tried to do this and got the following Mypy type check error: `core/controllers/voice_artist.py:101: error: Redundant cast to "AudioUploadHandlerNormalizedRequestDict"  [redundant-cast]`
The backend tests pass.

These are the changes I did:
Added RequestDict:
```python

class AudioUploadHandlerNormalizedRequestDict(TypedDict):
    """Dict representation of AudioUploadHandler's
    normalized_request dictionary.
    """

    raw_audio_file: bytes
```
Changed handler arguments:
```python
class AudioUploadHandler(
    base.BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict]
):
```

Cast in post method:
```python
        # Here we use cast because we want to narrow down the type
        # of 'normalized_request' to a particular TypedDict that
        # was defined according to the schemas so that the type
        # of fetched value is considered bytes and not str.
        request_data = cast(
            AudioUploadHandlerNormalizedRequestDict,
            self.normalized_request
        )
        raw_audio_file = request_data['raw_audio_file']
        filename = self.normalized_payload['filename']

```

sahiljoster32 replied:
explicit Cast in post method is not needed.

```python
        # Here we use cast because we want to narrow down the type
        # of 'normalized_request' to a particular TypedDict that
        # was defined according to the schemas so that the type
        # of fetched value is considered bytes and not str.
        request_data = cast(
            AudioUploadHandlerNormalizedRequestDict,
            self.normalized_request
        )
```
Now, you can just simply use 
```python
raw_audio_file = self.normalized_request['raw_audio_file']
```

sahiljoster32 replied:
The other two changes are good!

adamhalim replied:
Awesome, thanks! Just pushed the new changes

vojtechjelinek reviewed (approved):
LGTM!

vojtechjelinek reviewed (changes requested):
Left some comments.

adamhalim committed:
Merge branch 'audio-upload-handler-schema' of https://github.com/adamhalim/oppia into audio-upload-handler-schema

adamhalim committed:
Remove mutagen.mp3 import

adamhalim committed:
AssetDetHandlerAudioTest: add missing expected_status_int to all tests

adamhalim committed:
Merge branch 'develop' into audio-upload-handler-schema

adamhalim committed:
Merge branch 'audio-upload-handler-schema' of https://github.com/adamhalim/oppia into audio-upload-handler-schema

adamhalim committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audio-upload-handler-schema

adamhalim committed:
Add AudioUploadHandlerNormalizedRequestDict
In response to review comment

adamhalim committed:
Removed explicit cast

sahiljoster32 commented:
@vojtechjelinek PTAL!

vojtechjelinek reviewed (approved):
LGTM!
